hoomd/ParticleData.cc

```diff
@@ -1188,6 +1194,24 @@ void ParticleData::initializeFromSnapshot(const SnapshotParticleData<Real>& snap
 
     // notify listeners that number of types has changed
     m_num_types_signal.emit();
+
+    // Raise an exception if there are any invalid type ids. This is done here (instead of in the
+    // loops above) to avoid MPI communication deadlocks when only some ranks have invalid types.
+    // As a convenience, broadcast the value so that all ranks throw the exception.
+    #ifdef ENABLE_MPI
+    if (m_decomposition)
+        {
+        bcast(max_typeid, 0, m_exec_conf->getMPICommunicator());
+        }
+    #endif
+
+    if (max_typeid >= m_type_mapping.size())
```

Comment:
Is it valid to have a system with zero particles and zero types? If so, this would throw an error because `0 >= 0`. That might be undesired.

Reply:
It is valid to have a gsd file with 0 particles in a frame. I would need to check and see how HOOMD behaves with 0 particles in the whole system. My guess is most of the code will behave fine - it does need to support 0 particles on an individual MPI rank.

I would guess that most code in HOOMD assumes there is at least one particle type. Do you think it would be valid to run a simulation with 0 particle types? There would be no way to specify pair interaction potentials. I guess you could have 0 particle types but 1 bond type and model gaussian chains... I would have to test and see whether those code paths allow that and whether those that require at least 1 type raise appropriate errors.

Reply:
I regularly run HOOMD with no HOOMD particles and only MPCD particles. I removed a few checks on number of particles when I implemented MPCD to support this, but in general, everything seemed to work fine because ranks could already have zero particles.

I could see it being reasonable to have 0 particle types if and only if you have no HOOMD particles in a system. I don't think you can have a bond without particles, though, since bonds are defined between two particles; these would need to have a type.

Reply:
> I could see it being reasonable to have 0 particle types if and only if you have no HOOMD particles in a system. I don't think you can have a bond without particles, though, since bonds are defined between two particles; these would need to have a type.

This makes sense. I will update the check accordingly.